src/modules/flow/netctl/netctl.c

```diff
@@ -0,0 +1,227 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "sol-flow/netctl.h"
+#include "sol-flow-internal.h"
+
+#include <sol-netctl.h>
+#include <sol-str-table.h>
+#include <sol-util-internal.h>
+#include <errno.h>
+
+struct list_service_data {
+    struct sol_ptr_vector service_list;
+    struct sol_flow_node *node;
+};
+
+struct connection_service_data {
+    struct sol_flow_node *node;
+    char *ap_name;
+    bool connect;
+};
+
+void
+network_service_cb(void *data, const struct sol_netctl_service *service)
+{
+    struct list_service_data *mdata = data;
+    struct sol_flow_node *node = mdata->node;
+    int r;
+    const char *name;
+
+    if (sol_netctl_service_get_state(service) ==
+        SOL_NETCTL_SERVICE_STATE_REMOVE) {
+        sol_ptr_vector_del_element(&mdata->service_list, service);
+        return;
+    }
+
+    name = sol_netctl_service_get_name(service);
+    r = sol_flow_send_string_packet(node,
+        SOL_FLOW_NODE_TYPE_NETCTL_LIST_SERVICES__OUT__OUT,
+        name);
+    if (r < 0)
+        SOL_WRN("Failed to send network service name: %s", name);
+
+    sol_ptr_vector_append(&mdata->service_list, service);
+}
+
+void
+network_connection_cb(void *data, const struct sol_netctl_service *service)
+{
+    struct connection_service_data *mdata = data;
+    struct sol_flow_node *node = mdata->node;
+    enum sol_netctl_service_state state;
+    int r;
+    const char *name;
+
+    state = sol_netctl_service_get_state(service);
+    name = sol_netctl_service_get_name(service);
+
+    if (name && strcmp(name, mdata->ap_name) == 0)
+        if (state == SOL_NETCTL_SERVICE_STATE_IDLE && mdata->connect) {
+            sol_netctl_service_connect(service);
+            sol_flow_send_string_packet(node,
+                SOL_FLOW_NODE_TYPE_NETCTL_CONNECTION__OUT__OUT,
+                "Connect AP!");
+        } else if ((state == SOL_NETCTL_SERVICE_STATE_READY) && (!mdata->connect)) {
+            sol_netctl_service_disconnect(service);
+            sol_flow_send_string_packet(node,
+                SOL_FLOW_NODE_TYPE_NETCTL_CONNECTION__OUT__OUT,
+                "Disconnect AP!");
+        }
+}
```

Comment:
This function must be reworked if changing `CONNECT` and `DISCONNECT` ports to be a trigger (takes `any` packet type). This monitor would only be used to change the `OUT` port, that is to output the string representing the `sol_netctl_service_get_state()`, not `"Connect AP!"`  or `"Disconnect AP!"` (you'll need a `sol_netctl_service_state_to_str()`). As I've mentioned earlier, you may also have an `ONLINE` (boolean) port that will emit true/false if state is `SOL_NETCTL_STATE_ONLINE`. This is a nice helper.

In the `CONNECT` and `DISCONNECT` port process functions you'd call `sol_netctl_get_services()`, iterate over all services and see `streq(sol_netctl_service_get_name(itr), mdata->name)`. In this case you call `sol_netctl_service_connect(itr)` (or disconnect). If the iteration did not find any services, then it must emit an error packet `error = ENOENT, message = "Did not found service name %s", mdata->name`.